### PR TITLE
Update japicmp-gradle-plugin to 0.4.1

### PR DESCRIPTION
**Pull request by octylFractal**
This is in preparation for the expected merge of https://github.com/melix/japicmp-gradle-plugin/pull/57, which is likely to be released as 0.4.1.

~~This is in draft because there is a problem with fingerprinting which prevents tracking the state of the task. I'm not sure how to resolve it:~~
```
Caused by: java.lang.ClassNotFoundException: gradlebuild.binarycompatibility.rules.AcceptedRegressionsRulePostProcess
        at org.gradle.internal.io.ClassLoaderObjectInputStream.resolveClass(ClassLoaderObjectInputStream.java:40)
        at org.gradle.internal.snapshot.impl.JavaSerializedValueSnapshot.populateClass(JavaSerializedValueSnapshot.java:114)
        ... 71 more
```
~~This class is on the classpath when the task is originally configured; but it seems to get lost when loading the cached fingerprinting state.~~

This is not considered blocking, will be resolved after the merge.

### CODE REVIEW

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle
Comment (DPUkyle) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:139]: Would extracting a lambda parameter such as `report` cause the compiler to know this is an Action, not Closure, and allow you to remove the cast?

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy
Comment (DPUkyle) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:59]: Lots of eager resolution here; is that because the rule parameters like `publicApiPatterns` cannot accept a Provider?

Can we refactor `AbstractSuperClassChangesRule` like so? I'm not sure how we'd inject the ObjectFactory though.

```
-    private final List<Pattern> publicApiPatterns
+    private final ListProperty<Pattern> publicApiPatterns = objects.listProperty(Pattern)

```

**Review by DPUkyle — CHANGES_REQUESTED** [review]
LGTM with just a question or two.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle
Comment (blindpirate) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:139]: I think an extra lambda parameter is also a valid closure?

```
richReport({ it ->
    ...
} as Closure)
```

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy
Comment (blindpirate) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:59]: +1

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle
Comment (octylFractal) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild.binary-compatibility.gradle:139]: There's no `Closure` variant here. The only one is `Action`. I think this is a mistake on the plugin's part.

AFAIK, this is the only way to generate an `Action`. And it now uses `it` because Groovy doesn't do what Kotlin does with the `Action` interface and set the receiver to the parameter. I would've used `ConfigureUtil` here which makes more sense, but I was unsure if I should use an internal API like that, even in our scripts.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy
Comment (octylFractal) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:59]: Hmm, technically we're only supposed to be passing a `String` here, so even this is already suspicious (thanks Groovy for not checking generics...). I could skip using a `Property` and just use a `Provider`, since this isn't really assignable.

**Review by blindpirate — APPROVED** [review]

File: build-logic/build-platform/build.gradle.kts
Comment (DPUkyle) [inline build-logic/build-platform/build.gradle.kts:24]: Please bump to `0.4.1` and retest!

**Review by DPUkyle — CHANGES_REQUESTED** [review]

File: build-logic/build-platform/build.gradle.kts
Comment (octylFractal) [reply build-logic/build-platform/build.gradle.kts]: Done.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy
Comment (octylFractal) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:46]: Unless there is a disagreement, I am going to merge this after @DPUkyle approves, with this in place. Investigation will continue after this.

Comment (DPUkyle) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/BinaryCompatibilityHelper.groovy:46]: SGTM. @big-guy and I discussed offline and are fine if this task is not up-to-date/cacheable as it's _mostly_ invoked asynchronously on CI.

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Pull request merged**
